sdp/processors/tts/pyannote.py

```diff
@@ -0,0 +1,304 @@
+            # overlap starts during turn
+            turn_overlaps = True
+            break
+        elif (overlap.end <= turn.end) and (overlap.end >= turn.start):
+            # overlap ends during turn
+            turn_overlaps = True
+            break
+    return turn_overlaps
+
+class PyAnnoteDiarizationAndOverlapDetection(BaseProcessor):
+    """This processor performs speaker diarization and overlap detection using PyAnnote.
+
+    It processes audio files to identify different speakers and detect overlapping speech
+    segments using PyAnnote's speaker diarization pipeline and VAD (Voice Activity Detection).
+    The processor segments audio into speaker turns and identifies regions with overlapping speech.
+
+    Args:
+        hf_token (str): HuggingFace authentication token for accessing pretrained models
+        segmentation_batch_size (int, optional): Batch size for segmentation. Defaults to 128
+        embedding_batch_size (int, optional): Batch size for speaker embeddings. Defaults to 128
+        min_length (float, optional): Minimum length of segments in seconds. Defaults to 0.5
+        max_length (float, optional): Maximum length of segments in seconds. Defaults to 40
+        device (str, optional): Device to run the models on ('cuda' or 'cpu'). Defaults to "cuda"
+
+    Returns:
+        The same data as in the input manifest, but with speaker diarization and overlap
+        detection information added to each segment.
+
+    Example:
+        .. code-block:: yaml
+
+            - _target_: sdp.processors.tts.pyannote.PyAnnoteDiarizationAndOverlapDetection
+              input_manifest_file: ${workspace_dir}/manifest.json
+              output_manifest_file: ${workspace_dir}/manifest_diarized.json
+              hf_token: ${hf_token}
+    """
+
+    try:
+        import ndjson
+        from pyannote.audio import Pipeline
+        from pyannote.audio.pipelines.utils.hook import ProgressHook
+        from whisperx.audio import SAMPLE_RATE
+        from whisperx.vad import load_vad_model, merge_chunks
+        import torch
+        import torchaudio
+        _HAS_DEPS = True
+    except ImportError:
+        _HAS_DEPS = False
+    
+    def __init__(self,
+                 hf_token: str,
+                 segmentation_batch_size: int = 128,
+                 embedding_batch_size: int = 128,
+                 min_length: float = 0.5,
+                 max_length: float = 40,
+                 device: str = "cuda",
+                 **kwargs
+        ):
+
+        # only now we check for those imports
+        if not self._HAS_DEPS:
+             raise RuntimeError(
+                 "To use PyAnnoteDiarizationAndOverlapDetection you must install "
+                 "`pyannote.audio`, `torch`, `torchaudio` and `whisperx`."
+             )
+
+        super().__init__(**kwargs)
+
+
+        self.pipeline = Pipeline.from_pretrained("pyannote/speaker-diarization-3.1",
```

Comment:
here,
Second option is to upload this model to aws, and add a check if we can download it from aws/have it locally